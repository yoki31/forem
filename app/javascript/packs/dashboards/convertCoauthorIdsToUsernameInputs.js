import { h, render } from 'preact';
import { UsernameInput } from '@components/UsernameInput';
import { UserStore } from '@components/UserStore';
import '@utilities/document_ready';

async function fetchAllUsers(fetchUrls) {
  const users = await Promise.all(
    fetchUrls.map((fetchUrl) =>
      UserStore.fetch(fetchUrl).then((data) => [fetchUrl, data]),
    ),
  );
  return new Map(users);
}

function extractFetchUrl(field) {
  return field.dataset.fetchUsers;
}

function extractFetchUrls(fields) {
  const urls = [...fields].map((field) => extractFetchUrl(field));
  return [...new Set(urls)];
}

export async function convertCoauthorIdsToUsernameInputs() {
  const usernameFields = document.getElementsByClassName(
    'article_org_co_author_ids_list',
  );

  const fetchUrls = extractFetchUrls(usernameFields);
  const usersMap = await fetchAllUsers(fetchUrls);

  for (const targetField of usernameFields) {
    targetField.type = 'hidden';
    const exceptAuthorId =
      targetField.form.querySelector('#article_user_id').value;
    const inputId = `auto${targetField.id}`;

    const users = usersMap.get(extractFetchUrl(targetField));
    const row = targetField.parentElement;

    const value = users.matchingIds(targetField.value.split(','));
    const fetchSuggestions = function (term) {
      return users.search(term, { except: exceptAuthorId });
    };

    const handleSelectionsChanged = function (ids) {
      targetField.value = ids;
    };

    render(
      <UsernameInput
        labelText="Add up to 4"
        placeholder="Add up to 4..."
        maxSelections={4}
        inputId={inputId}
        defaultValue={value}
        fetchSuggestions={fetchSuggestions}
        handleSelectionsChanged={handleSelectionsChanged}
      />,
      row,
    );
  }
}

document.ready.then(() => {
  convertCoauthorIdsToUsernameInputs();
});
