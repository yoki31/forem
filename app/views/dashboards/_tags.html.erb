<%= javascript_include_tag "dashboardTags", defer: true %>

<div id="following-wrapper" class="grid gap-3 m:gap-4 s:grid-cols-2 l:grid-cols-3 px-3 m:px-0">
  <% tags.each do |follow| %>
    <% tag = follow.followable %>
    <% if tag %>
      <div class="dashboard__tag__container crayons-card p-3 pt-2 m:p-5 m:pt-4 relative flex flex-col single-article break-word content-center" id="follows-<%= follow.id %>" data-follow-id="<%= follow.id %>" data-tag-id="<%= tag.id %>">
        <div class="mb-1 flex items-center justify-between">
          <h4 class="-ml-2">
            <%= render_tag_link(tag.name) %>
          </h4>
          <div class="fs-xs color-base-60"><%= t("views.tags.published", count: number_with_delimiter(tag.taggings_count, delimiter: ",")) %></div>
        </div>
        <% if tag.short_summary.present? %>
          <p class="mb-6 fs-s color-base-70 truncate-at-3"><%= sanitize tag.short_summary %></p>
        <% end %>

        <div class="dashboard__tag__buttons-container">
          <% if page === "following" %>
            <div class="flex justify-between">
              <div>
                <button class="crayons-btn crayons-btn--outlined follow-button" aria-label="Following tag: <%= tag.name %>"><%= t("views.dashboard.tags.following_button") %></button>
              </div>
              <div class="dropdown-trigger-container relative">
                <button type="button" class="c-btn c-btn--icon-alone dropdown-trigger" id="options-dropdown-trigger-<%= tag.id %>" aria-haspopup="true" aria-expanded="false" aria-controls="options-dropdown">
                  <%= crayons_icon_tag("overflow-horizontal", title: t("views.dashboard.tags.options.icon")) %>
                </button>
                <div class="crayons-dropdown right-0 left-0 s:left-auto" id="options-dropdown-<%= tag.id %>">
                  <ul class="p-0">
                    <li>
                      <button type="button" class="hide-button c-btn w-100 align-left flex items-center" aria-label="Hide tag: <%= tag.name %>">
                        <%= crayons_icon_tag("eye-off", width: 18, height: 18, title: t("views.dashboard.tags.options.hide")) %> <span class="ml-2"><%= t("views.dashboard.tags.options.hide") %></span>
                      </button>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          <% end %>

          <% if page === "hidden" %>
            <div>
              <button class="c-btn c-btn--primary c-btn--destructive unhide-button" aria-label="Unhide tag: <%= tag.name %>"><%= t("views.dashboard.tags.unhide_button") %></button>
            </div>
          <% end %>
        </div>
      </div>
    <% end %>
  <% end %>
</div>
