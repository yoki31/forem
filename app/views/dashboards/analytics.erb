<% title t("views.dashboard.analytics.meta.title") %>

<main class="dashboard-container analytics-container crayons-layout" id="main-content">
  <nav aria-label="Dashboards" class="mx-5">
    <ul class="list-none">
      <li>
        <a href="<%= dashboard_path %>" class="rounded-btn crayons-link--block rounded-btn--transparent"><%= t("views.dashboard.analytics.back") %></a>
      </li>
      <% if @organizations.any? %>
        <li>
          <a class="rounded-btn crayons-link--block organization" 
             href="<%= dashboard_analytics_path %>"
             <%= params[:org_id] ? "" : 'aria-current="page"'.html_safe %>>
             <%= t("views.dashboard.analytics.your") %>
          </a>
        </li>
        <% @organizations.each do |org| %>
          <li>
            <a
              class="rounded-btn crayons-link--block organization"
              href="<%= dashboard_analytics_org_path(org.id) %>"
              data-organization-id="<%= org.id %>"
              <%= params[:org_id].to_i == org.id ? 'aria-current="page"'.html_safe : "" %>>
              <%= t("views.dashboard.analytics.org", org: org.name) %>
            </a>
          </li>
        <% end %>
      <% end %>
      </ul>
    </nav>
      

  <h1 class="fs-4xl fw-bold mx-5"><%= t("views.dashboard.analytics.heading", name: @user_or_org.name) %></h1>

  <%= render "shared/stats" %>
</main>

<%= javascript_include_tag "analyticsDashboard", defer: true %>
