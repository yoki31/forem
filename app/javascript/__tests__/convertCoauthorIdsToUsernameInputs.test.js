import fetch from 'jest-fetch-mock';
import '@testing-library/jest-dom';
import { userEvent } from '@testing-library/user-event';

import { convertCoauthorIdsToUsernameInputs } from '../packs/dashboards/convertCoauthorIdsToUsernameInputs';

global.fetch = fetch;

jest.mock('@utilities/debounceAction', () => ({
  debounceAction: fn => fn
}));

function fakeFetchResponseJSON() {
  return JSON.stringify([
    { name: 'Alice', username: 'alice', id: 1 },
    { name: 'Bob', username: 'bob', id: 2 },
    { name: 'Charlie', username: 'charlie', id: 3 },
  ]);
}

describe('convertCoauthorIdsToUsernameInputs', () => {
  beforeEach(() => {
    global.Honeybadger = { notify: jest.fn() };

    fetch.resetMocks();
    fetch.mockResponse(fakeFetchResponseJSON());
  });

  describe('when there is no pre-existing id value', () => {
    beforeEach(() => {
      window.document.body.innerHTML = `
      <form method="post">
        <div class="crayons-field mb-4">
          <label for="user_id">Author</label>
          <select class="crayons-select" name="article[user_id]" id="article_user_id">
            <option selected="selected" value="1">
              Alice
            </option>
            <option value="2">
              Bob
            </option>
            <option value="3">
              Charlie
            </option>
          </select>
        </div>

        <div class="crayons-field mb-4">
          <label for="co_author_ids">Co-authors</label>
          <div class="crayons-field">
            <input id="article_ID_co_author_ids_list"
                   class="article_org_co_author_ids_list"
                   name="article[co_author_ids_list]"
                   data-fetch-users="/org7053/members.json"
                   type="text"
                   value="" >
          </div>
        </div>

        <button type="submit" class="crayons-btn">Save</button>
      </form>
      `;
    });

    it('calls fetch, with exception for author ID', async () => {
      await convertCoauthorIdsToUsernameInputs();
      expect(fetch).toHaveBeenCalledWith('/org7053/members.json');
    });

    it('makes the co-author field hidden', async () => {
      await convertCoauthorIdsToUsernameInputs();
      const co_author_field = document.getElementById(
        'article_ID_co_author_ids_list',
      );
      expect(co_author_field.type).toBe('hidden');
    });

    it('renders matching snapshot', async () => {
      await convertCoauthorIdsToUsernameInputs();
      expect(document.forms[0].innerHTML).toMatchSnapshot();
    });

    it('works as expected', async () => {
      await convertCoauthorIdsToUsernameInputs();
      const input = document.querySelector(
        "input[placeholder='Add up to 4...']",
      );
      input.focus();
      await userEvent.type(input, 'Bob,');

      const hiddenField = document.querySelector(
        "input[name='article[co_author_ids_list]']",
      );
      expect(hiddenField.value).toBe('2');
    });
  });

  describe('when there *is* a pre-existing id value', () => {
    beforeEach(() => {
      window.document.body.innerHTML = `
      <form method="post">
        <div class="crayons-field mb-4">
          <label for="user_id">Author</label>
          <select class="crayons-select" name="article[user_id]" id="article_user_id">
            <option selected="selected" value="1">
              Alice
            </option>
            <option value="2">
              Bob
            </option>
            <option value="3">
              Charlie
            </option>
          </select>
        </div>

        <div class="crayons-field mb-4">
          <label for="co_author_ids">Co-authors</label>
          <div class="crayons-field">
            <input id="article_ID_co_author_ids_list"
                   class="article_org_co_author_ids_list"
                   name="article[co_author_ids_list]"
                   data-fetch-users="/org7053/members.json"
                   type="text"
                   value="3" >
          </div>
        </div>

        <button type="submit" class="crayons-btn">Save</button>
      </form>
      `;
    });

    it('renders matching snapshot', async () => {
      await convertCoauthorIdsToUsernameInputs();
      expect(document.forms[0].innerHTML).toMatchSnapshot();
    });

    it('works as expected', async () => {
      await convertCoauthorIdsToUsernameInputs();
      const input = document.querySelector(
        "input[placeholder='Add another...']",
      );
      input.focus();
      await userEvent.type(input, 'Bob,');

      const hiddenField = document.querySelector(
        "input[name='article[co_author_ids_list]']",
      );
      expect(hiddenField.value).toBe('3, 2');
    });

    it('can remove previously selected', async () => {
      await convertCoauthorIdsToUsernameInputs();
      const deselect = document.querySelector(
        '.c-autocomplete--multi__selected',
      );
      await userEvent.click(deselect);

      const hiddenField = document.querySelector(
        "input[name='article[co_author_ids_list]']",
      );
      expect(hiddenField.value).toBe('');
    });
  });
});
