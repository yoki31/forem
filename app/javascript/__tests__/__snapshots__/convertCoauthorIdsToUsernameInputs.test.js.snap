// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertCoauthorIdsToUsernameInputs when there *is* a pre-existing id value renders matching snapshot 1`] = `
"
        <div class=\\"crayons-field mb-4\\">
          <label for=\\"user_id\\">Author</label>
          <select class=\\"crayons-select\\" name=\\"article[user_id]\\" id=\\"article_user_id\\">
            <option selected=\\"selected\\" value=\\"1\\">
              Alice
            </option>
            <option value=\\"2\\">
              Bob
            </option>
            <option value=\\"3\\">
              Charlie
            </option>
          </select>
        </div>

        <div class=\\"crayons-field mb-4\\">
          <label for=\\"co_author_ids\\">Co-authors</label>
          <div class=\\"crayons-field\\"><span aria-hidden=\\"true\\" class=\\"absolute pointer-events-none opacity-0 p-2\\"></span><label id=\\"multi-select-label\\" class=\\"screen-reader-only\\">Add up to 4</label><span id=\\"input-description\\" class=\\"screen-reader-only\\">Maximum 4 selections</span><div class=\\"screen-reader-only\\"><p>Selected items:</p><ul aria-live=\\"assertive\\" aria-atomic=\\"false\\" aria-relevant=\\"additions removals\\" class=\\"screen-reader-only list-none\\"></ul></div><div class=\\"c-autocomplete--multi relative\\"><div role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-expanded=\\"false\\" aria-owns=\\"listbox1\\" class=\\"c-autocomplete--multi__wrapper-border crayons-textfield flex items-center  cursor-text\\"><ul id=\\"combo-selected\\" class=\\"list-none flex flex-wrap w-100\\"><li style=\\"order: 1;\\" class=\\"c-autocomplete--multi__selection-list-item w-max\\"><div role=\\"group\\" aria-label=\\"Charlie\\" class=\\"flex mr-1 mb-1 w-max\\"><button type=\\"button\\" aria-label=\\"Edit Charlie\\" class=\\"c-btn c-btn--secondary c-autocomplete--multi__selected p-1 cursor-text\\">Charlie</button><button type=\\"button\\" aria-label=\\"Remove Charlie\\" class=\\"c-btn c-btn--secondary c-autocomplete--multi__selected p-1\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"crayons-icon\\"><path d=\\"m12 10.586 4.95-4.95 1.414 1.414-4.95 4.95 4.95 4.95-1.414 1.414-4.95-4.95-4.95 4.95-1.414-1.414 4.95-4.95-4.95-4.95L7.05 5.636l4.95 4.95z\\"></path></svg></button></div></li><li style=\\"order: 2;\\" class=\\"self-center\\"><input id=\\"autoarticle_ID_co_author_ids_list\\" autocomplete=\\"off\\" aria-autocomplete=\\"list\\" aria-labelledby=\\"multi-select-label selected-items-list\\" aria-describedby=\\"input-description\\" aria-disabled=\\"false\\" type=\\"text\\" placeholder=\\"Add another...\\" class=\\"c-autocomplete--multi__input\\"></li></ul></div></div>
            <input id=\\"article_ID_co_author_ids_list\\" class=\\"article_org_co_author_ids_list\\" name=\\"article[co_author_ids_list]\\" data-fetch-users=\\"/org7053/members.json\\" type=\\"hidden\\" value=\\"3\\">
          </div>
        </div>

        <button type=\\"submit\\" class=\\"crayons-btn\\">Save</button>
      "
`;

exports[`convertCoauthorIdsToUsernameInputs when there is no pre-existing id value renders matching snapshot 1`] = `
"
        <div class=\\"crayons-field mb-4\\">
          <label for=\\"user_id\\">Author</label>
          <select class=\\"crayons-select\\" name=\\"article[user_id]\\" id=\\"article_user_id\\">
            <option selected=\\"selected\\" value=\\"1\\">
              Alice
            </option>
            <option value=\\"2\\">
              Bob
            </option>
            <option value=\\"3\\">
              Charlie
            </option>
          </select>
        </div>

        <div class=\\"crayons-field mb-4\\">
          <label for=\\"co_author_ids\\">Co-authors</label>
          <div class=\\"crayons-field\\"><span aria-hidden=\\"true\\" class=\\"absolute pointer-events-none opacity-0 p-2\\"></span><label id=\\"multi-select-label\\" class=\\"screen-reader-only\\">Add up to 4</label><span id=\\"input-description\\" class=\\"screen-reader-only\\">Maximum 4 selections</span><div class=\\"screen-reader-only\\"><p>Selected items:</p><ul aria-live=\\"assertive\\" aria-atomic=\\"false\\" aria-relevant=\\"additions removals\\" class=\\"screen-reader-only list-none\\"></ul></div><div class=\\"c-autocomplete--multi relative\\"><div role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-expanded=\\"false\\" aria-owns=\\"listbox1\\" class=\\"c-autocomplete--multi__wrapper-border crayons-textfield flex items-center  cursor-text\\"><ul id=\\"combo-selected\\" class=\\"list-none flex flex-wrap w-100\\"><li style=\\"order: 1;\\" class=\\"self-center\\"><input id=\\"autoarticle_ID_co_author_ids_list\\" autocomplete=\\"off\\" aria-autocomplete=\\"list\\" aria-labelledby=\\"multi-select-label selected-items-list\\" aria-describedby=\\"input-description\\" aria-disabled=\\"false\\" type=\\"text\\" placeholder=\\"Add up to 4...\\" class=\\"c-autocomplete--multi__input\\"></li></ul></div></div>
            <input id=\\"article_ID_co_author_ids_list\\" class=\\"article_org_co_author_ids_list\\" name=\\"article[co_author_ids_list]\\" data-fetch-users=\\"/org7053/members.json\\" type=\\"hidden\\" value=\\"\\">
          </div>
        </div>

        <button type=\\"submit\\" class=\\"crayons-btn\\">Save</button>
      "
`;
